src/completions/generate-v1.ts

```diff
@@ -0,0 +1,277 @@
+
+import { homedir } from 'os';
+
+const completionsVersion = 1;
+
+interface TagDictionary {
+	[key: string]: Tag;
+}
+interface Tag {
+	apiName: string;
+}
+
+interface PropertiesDictionary {
+	[key: string]: Property;
+}
+interface Property {
+	description: string;
+	type: string;
+	values?: string[];
+}
+
+interface TypeDictionary {
+	[key: string]: Type;
+}
+interface Type {
+	description: string;
+	events: string[];
+	functions: string[];
+	properties: string[];
+}
+
+/**
+ * Generate completions for an Alloy version.
+ *
+ * @param {Boolean} force - Force generation of completion file.
+ */
+export async function generateAlloyCompletions (force: boolean) {
+	const appcPath = path.join(homedir(), '.appcelerator', 'install');
+	const version = await checkInstalledVersion();
+	if (!version) {
+		throw Error('Unable to find installed alloy version.');
+	}
+	const alloyPath = path.join(appcPath, version, 'package', 'node_modules', 'alloy');
+	const alloyVersion = await getAlloyVersion();
+
+	const alloyCompletionsFilename = getAlloyCompletionsFileName(alloyVersion, completionsVersion);
+
+	if (!force && await fs.pathExists(alloyCompletionsFilename)) {
+		return;
+	}
+
+	// TODO: Generate completions from this?
+	// const alloyApi = await fs.readJSON(path.join(alloyPath, 'docs', 'api.jsca'));
+
+	// generate tag list
+	const alloyTags = await fs.readdir(path.join(alloyPath, 'Alloy', 'commands', 'compile', 'parsers'));
+	const tagDic: TagDictionary = {};
+	for (const tag of alloyTags) {
+		if (!tag.endsWith('.js')) {
+			continue;
+		}
+		const ar = tag.split('.');
+		const tagName = ar[ar.length - 2];
+		if (tagName.indexOf('_') !== 0 && tagName[0] === tagName[0].toUpperCase()) {
+			tagDic[tagName] = {
+				apiName: tag.replace('.js', '')
+			};
+		} else if (tagName === '_ProxyProperty' && tag.indexOf('Ti.UI') === 0) {
+			tagDic[ar[ar.length - 3]] = { // Ti.UI.Window._ProxyProperty
+				apiName: tag.replace('.js', '').replace('._ProxyProperty', '')
+			};
+		}
+	}
+
+	// add missing tags
+	Object.assign(tagDic, {
+		View: {
+			apiName: 'Ti.UI.View'
+		},
+		Templates: {},
+		HeaderView: {},
+		FooterView: {},
+		ScrollView: {
+			apiName: 'Ti.UI.ScrollView'
+		},
+		Slider: {
+			apiName: 'Ti.UI.Slider'
+		},
+		TableViewRow: {
+			apiName: 'Ti.UI.TableViewRow'
+		},
+		Alloy: {},
+		ActivityIndicator: {
+			apiName: 'Ti.UI.ActivityIndicator'
+		},
+		WebView: {
+			apiName: 'Ti.UI.WebView'
+		}
+	});
+
+	const sortedTagDic: TagDictionary = {};
+	Object.keys(tagDic)
+		.sort()
+		.forEach(k => sortedTagDic[k] = tagDic[k]);
+	try {
+		await fs.ensureDir(path.dirname(alloyCompletionsFilename));
+		await fs.writeJSON(alloyCompletionsFilename, {
+			version: 1,
+			alloyVersion,
+			tags: sortedTagDic
+		},
+		{
+			spaces: '\t'
+		});
+		return alloyVersion;
+	} catch (error) {
+		throw error;
+	}
+}
+
+/**
+ *
+ * Generate completions file for a Titanium SDK.
+ *
+ * @param {Boolean} [force=false] - Force generation of the completion file. 
```

Comment:
Force doesn't look to be optional anymore so please drop the square brackets in the jsdoc